src/sources/aws_sqs/source.rs

```diff
@@ -90,47 +90,47 @@ impl SqsSource {
         };
 
         if let Some(messages) = receive_message_output.messages {
-            let mut receipts_to_ack = vec![];
+            let mut receipts_to_ack = Vec::with_capacity(messages.len());
+            let mut events = Vec::with_capacity(messages.len());
+            let mut byte_size = 0;
 
             let (batch, batch_receiver) = BatchNotifier::maybe_new_with_receiver(acknowledgements);
             for message in messages {
                 if let Some(body) = message.body {
-                    emit!(&AwsSqsBytesReceived {
-                        byte_size: body.len()
-                    });
+                    byte_size += body.len();
+                    // a receipt handle should always exist
+                    if let Some(receipt_handle) = message.receipt_handle {
+                        receipts_to_ack.push(receipt_handle);
+                    }
                     let timestamp = get_timestamp(&message.attributes);
-                    let events = decode_message(self.decoder.clone(), body.as_bytes(), timestamp);
-                    let send_result = if let Some(batch) = batch.as_ref() {
-                        let events = events.map(|event| event.with_batch_notifier(batch));
-                        out.send_batch(events).await
+                    let decoded = decode_message(self.decoder.clone(), body.as_bytes(), timestamp);
+                    if let Some(batch) = batch.as_ref() {
+                        let decoded = decoded.map(|event| event.with_batch_notifier(batch));
+                        events.extend(decoded);
                     } else {
-                        out.send_batch(events).await
-                    };
-
-                    match send_result {
-                        Err(err) => error!(message = "Error sending to sink.", error = %err),
-                        Ok(()) => {
-                            // a receipt handle should always exist
-                            if let Some(receipt_handle) = message.receipt_handle {
-                                receipts_to_ack.push(receipt_handle);
-                            }
-                        }
+                        events.extend(decoded);
                     }
                 }
             }
             drop(batch);
```

Comment:
This may not be relevant, but why do we drop the batch explicitly?

Reply:
To ensure there isn't an extraneous reference to the batch, which would deadlock acknowledgement handling.

Reply:
@jszwedko will probably not the last one to stumble on that, I'd suggest to leave a comment for the next person now that we have the context 👍😄